[choreolib, docs] ChoreoLib docs overhaul (#994)
SleipnirGroup/Choreo: docs/choreolib/trajectory-api.md

# Trajectory API

!!! tip
    Java users are recommended to use an [Auto Factory](./auto-factory.md) instead of the Trajectory API for loading and following trajectories, as it provides a more user-friendly experience for developing complex autonomous routines.

To begin utilizing trajectories created by Choreo, you must first load a `Trajectory` ([Java](/api/choreolib/java/choreo/trajectory/Trajectory.html), [C++](/api/choreolib/cpp/classchoreo_1_1Trajectory.html)) from its corresponding file.

=== "Java"

    ```java
    // Loads from deploy/choreo/myTrajectory.traj
    // Optional.empty() is returned if the file does not exist or is invalid
    var trajectory = Choreo.loadTrajectory("myTrajectory");
    ```

=== "C++"

    ```cpp
    // Loads from deploy/choreo/myTrajectory.traj
    // std::nullopt is returned if the file does not exist or is invalid

>     auto trajectory = choreo::Choreo::LoadTrajectory<choreo::SwerveSample>("myTrajectory") // (1)

Reviewer: ```suggestion
    auto trajectory = choreo::Choreo::LoadTrajectory<choreo::SwerveSample>("myTrajectory"); // (1)
```

                if (initialPose.isPresent()) {
                    // Reset odometry to the start of the trajectory
                    driveSubsystem.resetOdometry(initialPose.get());
                }
            }

            // Reset and start the timer when the autonomous period begins
            timer.restart();
        }

        @Override
        public void autonomousPeriodic() {
            if (trajectory.isPresent()) {
                // Sample the trajectory at the current time into the autonomous period
                Optional<SwerveSample> sample = trajectory.get().sampleAt(timer.get(), isRedAlliance());

                if (sample.isPresent()) {
                    driveSubsystem.followTrajectory(sample);
                }
            }
        }

        private boolean isRedAlliance() {
            return DriverStation.getAlliance().orElse(Alliance.Blue).equals(Alliance.Red);
        }
    }
    ```

=== "C++"

    === "Source"

        ```cpp title="Robot.cpp"
        void Robot::AutonomousInit() {
            if (trajectory.has_value()) {
                // Get the initial pose of the trajectory

>                 std::optional<frc::Pose2d> initialPose = trajectory.value().GetInitialPose(IsRedAlliance());
> 
>                 if (initialPose.has_value()) {
>                     // Reset odometry to the start of the trajectory
>                     driveSubsystem.ResetOdometry(initialPose.value());

Reviewer: ```suggestion
                if (auto initialPose = trajectory.value().GetInitialPose(IsRedAlliance())) {
                // Reset odometry to the start of the trajectory
                driveSubsystem.ResetOdometry(initialPose.value());
```

                }
            }

            // Reset and start the timer when the autonomous period begins
            timer.Restart();
        }

        void Robot::AutonomousPeriodic() {
            if (trajectory.has_value()) {
                // Sample the trajectory at the current time into the autonomous period

>                 std::optional<choreo::SwerveSample> sample = trajectory.value().SampleAt(timer.Get(), IsRedAlliance());
> 
>                 if (sample.has_value()) {
>                     driveSubsystem.FollowTrajectory(sample.value());
>                 }

Reviewer: ```suggestion
                if (auto sample = trajectory.value().SampleAt(timer.Get(), IsRedAlliance())) {
                driveSubsystem.FollowTrajectory(sample.value());
                }
```

                    choreo::Choreo::LoadTrajectory<choreo::SwerveSample>("myTrajectory");

                Drive drive;
                frc::Timer timer;
        };
        ```

=== "Python"

    ```py
    class MyRobot(wpilib.TimedRobot):
        def robotInit(self):
            # Loads a swerve trajectory, alternatively use load_differential_trajectory if the robot is tank drive
            try:
                self.trajectory = choreo.load_swerve_trajectory("myTrajectory") # (1)
            except ValueError:
                self.trajectory = None

            self.drive_subsystem = Drive()
            self.timer = wpilib.Timer()

        def autonomousInit(self):
            if self.trajectory is not None:
                # Get the initial pose of the trajectory
                initial_pose = self.trajectory.get_initial_pose(self.is_red_alliance())

                if initial_pose is not None:
                    # Reset odometry to the start of the trajectory
                    self.drive_subsystem.reset_odometry(initial_pose)

            # Reset and start the timer when the autonomous period begins
            self.timer.restart()

        def autonomousPeriodic(self):

>             if self.trajectory is not None:

Reviewer: `is not None` is redundant. `if self.trajectory:` is sufficient.

To begin utilizing trajectories created by Choreo, you must first load a `Trajectory` ([Java](/api/choreolib/java/choreo/trajectory/Trajectory.html), [C++](/api/choreolib/cpp/classchoreo_1_1Trajectory.html)) from its corresponding file.

=== "Java"

    ```java
    // Loads from deploy/choreo/myTrajectory.traj
    // Optional.empty() is returned if the file does not exist or is invalid
    var trajectory = Choreo.loadTrajectory("myTrajectory");
    ```

=== "C++"

    ```cpp
    // Loads from deploy/choreo/myTrajectory.traj
    // std::nullopt is returned if the file does not exist or is invalid
    auto trajectory = choreo::LoadTrajectory("myTrajectory")
    ```

=== "Python"

    ```py
    # Loads from deploy/choreo/myTrajectory.traj
    # ValueError is thrown if the file does not exist or is invalid
    try:
      trajectory = choreo.load_swerve_trajectory("myTrajectory")
    except ValueError:
      # If the trajectory is not found, ChoreoLib already prints to DriverStation
      pass
    ```

!!! warning

>     Trajectories should always be loaded at startup, not when the autonomous period begins. Loading trajectories is a blocking operation, and larger trajectories may take multiple seconds to load on a RoboRIO, cutting into the time a robot has to run its autonomous routine in a match.

Reviewer 2: We should note that the Choreo AutoChooser's behaviour. The AutoChooser loads on select.
